src/coreclr/jit/morph.cpp

```diff
@@ -13160,6 +13148,57 @@ GenTree* Compiler::fgMorphMultiOp(GenTreeMultiOp* multiOp)
 }
 #endif // defined(FEATURE_SIMD) || defined(FEATURE_HW_INTRINSICS)
 
+//------------------------------------------------------------------------
+// fgMorphModToZero: Transform 'a % 1' or 'a % -1' into the equivalent '0'.
+//
+// Arguments:
+//    tree - The GT_MOD/GT_UMOD tree to morph
+//
+// Returns:
+//    The morphed tree, will be a GT_COMMA or a zero constant node.
+//
+GenTree* Compiler::fgMorphModToZero(GenTreeOp* tree)
+{
+    JITDUMP("\nMorphing MOD/UMOD [%06u] to Zero\n", dspTreeID(tree));
+
+    assert(tree->OperIs(GT_MOD, GT_UMOD));
+    assert(tree->gtOp2->IsIntegralConst(1) || tree->gtOp2->IsIntegralConst(-1));
+
+    GenTree* op1 = tree->gtGetOp1();
+    GenTree* op2 = tree->gtGetOp2();
+
+    const var_types type = tree->TypeGet();
+
+    GenTree* zero = gtNewZeroConNode(type);
```

Comment:
TP nit: it would be better to reuse an existing constant node (`op2`) instead of creating a new one.

Reply:
We shouldn't need to update for VN should we?

Reply:
Why not?

(Of at least two trees - the comma and the zero)

Reply:
Updated the VN and re-using `op2`. Not sure how to update the comma for VN.

Reply:
> Not sure how to update the comma for VN.

It's a bit non-trivial. The comma needs to have the normal VN of zero and `op1SideEffects`'s exception set.

We usually shy away from complex VN updates like this in morph and guard the transform on `fgGlobalMorph` instead.

Reply:
Do you suggest we not do the transformation if `op1` has side effects and it's not `fgGlobalMorph`? I guess I'm a little confused because there are cases where we create GT_COMMAs without updating VNs.

Reply:
It's not updating the VN, but maybe `gtUpdateNodeSideEffects` is sufficient enough?

Reply:
> Do you suggest we not do the transformation if op1 has side effects and it's not `fgGlobalMorph`

There are multiple ways to solve this problem.

1) The simplest (but most pessimistic) is to only do the optimization under `fgGlobalMorph` and not worry about VN updates.

2) A less pessimistic one would be to do the part of the transformation that doesn't require VN complex updates always, and one that does - under `fgGlobalMorph`. This would be differentiated by `op1->gtFlags & GTF_SIDE_EFFECT`.

3) Finally, you could introduce a method, call it `fgUpdateCommaValueNumber` that will do the appropriate thing and call it here.

So far we've been doing `1` in morph, but if the diffs for this change mean it's not useful if not done with `2` or `3`, I guess we can do those.

> there are cases where we create GT_COMMAs without updating VNs 

In general, any code that creates / changes nodes without paying attention to VNs after VNs have been established and before they are invalidated (~rationalization) is bugged.

Reply:
If you look at `fgMorphModToSubMulDiv`, we create new `GT_COMMA`s there but it's not guarded by `fgGlobalMorph` - does that mean it's bugged as well?

Reply:
Yes in theory, no in practice, because we nothing introduces new `MOD/UMOD` nodes after global morph, and the ones that were there would all be transformed. There are a few other "fixed point" transformations like that (e. g. block morphing).

Reply:
Majority of the optimization still works if we only do it in global morph when there are side effects, so it's sufficient.